### PR TITLE
Add deployment configuration for Stun server

### PR DESCRIPTION
**Pull request by klejejs**

### CODE REVIEW

File: stun_server/main.tf [hunk 1]
```diff
@@ -0,0 +1,18 @@
+terraform {
+  cloud {
+    organization = "home_assistant"
+
+    workspaces {
+      name = "infrastructure"
```
Comment (ludeeus) [inline stun_server/main.tf]:
```suggestion
      name = "stun_server"
```

File: stun_server/main.tf [hunk 2]
```diff
@@ -0,0 +1,18 @@
+terraform {
```
Comment (ludeeus) [inline stun_server/main.tf:1]: Please reuse webservice (or service) from the .modules subdirectory.

File: .modules/webservice/dns.tf [hunk 3]
```diff
@@ -5,8 +5,8 @@ data "cloudflare_zone" "dns_zone" {
 resource "cloudflare_record" "instance_dns" {
   zone_id = data.cloudflare_zone.dns_zone.id
   name    = coalesce(var.subdomain, lower(var.service_name))
-  value   = lower(aws_alb.main.dns_name)
+  content = lower(aws_alb.main.dns_name)
```
Comment (ludeeus) [inline .modules/webservice/dns.tf]: Unrelated change.
Move this to a separate PR (it also needs to update the provider).

File: .modules/service/ecs.tf [hunk 4]
```diff
@@ -56,4 +56,4 @@ resource "aws_ecs_task_definition" "task" {
       }
     }, var.container_definitions)
   ])
-}
\ No newline at end of file
+}
```
Comment (ludeeus) [inline .modules/service/ecs.tf]: Revert this change.

File: .modules/webservice/dns.tf (hunk 3 above)
Comment (klejejs) [reply .modules/webservice/dns.tf]: OK, will do that.

File: .modules/service/ecs.tf (hunk 4 above)
Comment (klejejs) [reply .modules/service/ecs.tf]: Reverted. But as it was auto-applied by the formatter, I think it should be modified in the future.

File: stun_server/region/ecs.tf [hunk 5]
```diff
@@ -0,0 +1,30 @@
+resource "aws_ecs_service" "stun-server" {
```
Comment (ludeeus) [inline stun_server/region/ecs.tf:1]: <https://github.com/home-assistant/deployments/pull/87/files#diff-53a630ff67f218bf174fff7b35b58537507d636383623b0088562416f92eb384R12>
Service is also defined here, so this is conflicting.

File: stun_server/region/module.tf [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+data "tfe_outputs" "infrastructure" {
+  organization = "home_assistant"
+  workspace    = "infrastructure"
+}
+
+module "stun_server" {
+  source = "../../.modules/service"
+
+  service_name      = local.service_name
+  container_image   = "ghcr.io/home-assistant/stun-server"
```
Comment (ludeeus) [inline stun_server/region/module.tf]: https://github.com/home-assistant/core/pkgs/container/stun-server returns 404

File: stun_server/region/module.tf [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+  container_version = var.image_tag
+  launch_type       = local.launch_type
```
Comment (ludeeus) [inline stun_server/region/module.tf]: fargate is the deafult, no need for this

File: stun_server/region/module.tf [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+  region            = var.region
```
Comment (ludeeus) [inline stun_server/region/module.tf]: use the current region data source instead

File: stun_server/main.tf [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+provider "aws" {
+  region = "us-east-1"
+}
+
+module "us_east_1" {
+  source = "./region"
+
+  region      = "us-east-1"
+  domain_name = var.domain_name
+  subdomain   = "stun-us"
```
Comment (ludeeus) [inline stun_server/main.tf]: lets concat these with the full region names (can even be handled automatically in the dns.tf file)

File: stun_server/main.tf [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+    }
+  }
+}
+
+provider "aws" {
+  region = "us-east-1"
+}
+
+module "us_east_1" {
+  source = "./region"
```
Comment (ludeeus) [inline stun_server/main.tf:16]: module is missing aws config, so us-east-1 will be used.

**Review by ludeeus — CHANGES_REQUESTED** [review]

File: stun_server/region/module.tf (hunk 6 above)
Comment (klejejs) [reply stun_server/region/module.tf]: I made a release and `ghcr.io/home-assistant/stun:1.0.0` returns an image now. Your provided URL still returns 404, but I don't think it is ever used in this context.

File: stun_server/region/module.tf (hunk 8 above)
Comment (klejejs) [reply stun_server/region/module.tf]: Changed. Also, used the region data source in other places as well.

File: stun_server/main.tf (hunk 10 above)
Comment (klejejs) [reply stun_server/main.tf:16]: Added.

File: stun_server/main.tf (hunk 9 above)
Comment (klejejs) [reply stun_server/main.tf]: Do you really think the URLs should include the full region in the subdomain? IMO, it makes them unnecessarily longer and adds unneeded information.

File: stun_server/region/module.tf (hunk 6 above)
Comment (ludeeus) [reply stun_server/region/module.tf]: As the requested image here is stun-server and not stun, yes it is used in this context.

File: stun_server/main.tf (hunk 9 above)
Comment (ludeeus) [reply stun_server/main.tf]: Yes, while they are not needed now, we might want more locations in the same continent later

File: stun_server/region/module.tf (hunk 6 above)
Comment (klejejs) [reply stun_server/region/module.tf]: 🤡

File: stun_server/main.tf (hunk 9 above)
Comment (klejejs) [reply stun_server/main.tf]: All right, if you think there will ever be a need.

File: stun_server/outputs.tf [hunk 11]
```diff
@@ -0,0 +1,14 @@
+output "stun_server_ip-us_east_1" {
```
Comment (ludeeus) [inline stun_server/outputs.tf]: Lets do 1 output for "ip" that is a dictionary containing all of them.

File: stun_server/region/module.tf [hunk 12]
```diff
@@ -0,0 +1,42 @@
+locals {
+  service_name = "stun-server"
+  launch_type  = "FARGATE"
```
Comment (ludeeus) [inline stun_server/region/module.tf]: Only used a single place, no need for a local

File: stun_server/main.tf [hunk 13]
```diff
@@ -0,0 +1,40 @@
...
+  subdomain   = "stun"
+  image_tag   = var.image_tag
+}
+
+module "ap_southeast_1" {
+  source = "./region"
+
+  region      = "ap-southeast-1"
+  domain_name = var.domain_name
+  subdomain   = "stun"
```
Comment (ludeeus) [inline stun_server/main.tf]: Does not really add much as this value is not expected to be dynamic, so move this to the dns.tf file directly instead of using variables.

File: stun_server/region/module.tf [hunk 14]
```diff
@@ -0,0 +1,42 @@
+locals {
```
Comment (ludeeus) [inline stun_server/region/module.tf:1]: `data.tfe_outputs.infrastructure.values[data.aws_region.current.name]` is reused quite a fair bit, might be a contender to be a local as `infrastructure_regioin`

Comment (klejejs) [reply stun_server/region/module.tf:1]: Made a local `infrastructure_region_outputs`.

File: stun_server/outputs.tf (hunk 11 above)
Comment (klejejs) [reply stun_server/outputs.tf]: Created an output `stun_server_ip` with all IPs.

Comment (ludeeus) [reply stun_server/outputs.tf]: They are all “stun_server” that is what this workspace is, prefixing the output with that is redundant.

Comment (klejejs) [reply stun_server/outputs.tf]: Fine. 👍🏻

**Review by bemble — DISMISSED** [review]
OK for me

File: stun_server/region/ecs.tf [hunk 15]
```diff
@@ -0,0 +1,34 @@
...
+  enable_ecs_managed_tags = true
+  wait_for_steady_state   = true
+
+  network_configuration {
+    assign_public_ip = true
+    security_groups  = [aws_security_group.stun_sg.id]
+    subnets = [
+      local.infrastructure_region_outputs.public_subnets[0],
+      local.infrastructure_region_outputs.public_subnets[1]
+    ]
```
Comment (ludeeus) [inline stun_server/region/ecs.tf]:
```suggestion
    subnets = local.infrastructure_region_outputs.public_subnets
```
Wouldnt that result in the same?

File: stun_server/region/module.tf [hunk 16]
```diff
@@ -0,0 +1,42 @@
...
+
+module "stun_server" {
+  source = "../../.modules/service"
+
+  service_name      = local.service_name
+  container_image   = "ghcr.io/home-assistant/stun"
+  container_version = var.image_tag
+  region            = data.aws_region.current.name
+  ecs_cpu           = 2048
+  ecs_memory        = 4096
```
Comment (ludeeus) [inline stun_server/region/module.tf]: This is quite big, do we need that?

File: stun_server/region/module.tf [hunk 17]
```diff
@@ -0,0 +1,42 @@
...
+      },
+      {
+        containerPort = 3478
+        hostPort      = 3478
+        protocol      = "udp"
+      }
+    ],
+  }
+  webservice      = true
+  rolling_updates = true
```
Comment (ludeeus) [inline stun_server/region/module.tf]: This is only used if the module is setting up the service, which is not the case here so it should be removed.

File: stun_server/region/dns.tf [hunk 18]
```diff
@@ -0,0 +1,11 @@
+data "cloudflare_zone" "dns_zone" {
+  name = var.domain_name
+}
+
+resource "cloudflare_record" "instance_dns" {
+  zone_id = data.cloudflare_zone.dns_zone.id
+  name    = join("-", ["stun", data.aws_region.current.name])
+  content = data.aws_network_interface.stun_server_interface.association[0].public_ip
```
Comment (ludeeus) [inline stun_server/region/dns.tf]: How does this handle rolling updates?
How does this work with new deployments?
stun_server_interface does a filter, right now, that will return nothing?
During a rolling update it will return old values?
Do we need to adjust the dependency here to ensure this resource is updated after the new task is active?

Comment (klejejs) [reply stun_server/region/dns.tf]: This needs to be tested after the merge. In theory, it should filter after ECS is deployed due to this config:
```
# Required to fetch the public IP address of the ECS service
  enable_ecs_managed_tags = true
  wait_for_steady_state   = true
```

Comment (ludeeus) [reply stun_server/region/dns.tf]: Even with `depends_on` being blank?

File: stun_server/region/ecs.tf (hunk 15 above)
Comment (klejejs) [reply stun_server/region/ecs.tf]: Somewhere, at some point, I used this and it did not work as expected. Does. not seem to be the case here.

File: stun_server/region/module.tf (hunk 16 above)
Comment (klejejs) [reply stun_server/region/module.tf]: Decreased, as we discussed.

File: stun_server/region/dns.tf (hunk 18 above)
Comment (klejejs) [reply stun_server/region/dns.tf]: Added `depends_on` as we discussed.

**Review by bemble — APPROVED** [review]
We just need to check if the dependency between the DNS record and the interface is enough (what might happened is the network interface is created and attached, so TF moves on, but there's no public IP yet).

**Review by ludeeus — APPROVED** [review]
